### PR TITLE
wen-restart: Find heaviest fork

### PR DESCRIPTION
**Pull request by wen-coding**
- Pass final result of aggregate RestartLastVotedForkSlots to next stage
- Find the heaviest fork based on the information

### CODE REVIEW

File: wen-restart/src/wen_restart.rs [hunk 1]
```diff
@@ -37,9 +40,14 @@ use {
 
 // If >42% of the validators have this block, repair this block locally.
 const REPAIR_THRESHOLD: f64 = 0.42;
+// When counting Heaviest Fork, only count those > active_stake - 38% * total_stake.
+const HEAVIEST_FORK_THRESHOLD: f64 = 0.38;
```
Comment (carllin) [inline wen-restart/src/wen_restart.rs]: will this always be equivalent to the switching proof threshold?

File: wen-restart/src/wen_restart.rs [hunk 2]
```diff
@@ -50,6 +58,28 @@ pub enum WenRestartError {
 impl std::fmt::Display for WenRestartError {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         match self {
+            WenRestartError::BlockNotFound(slot) => {
+                write!(f, "Block should be repaired but not found: {}", slot)
```
Comment (carllin) [inline wen-restart/src/wen_restart.rs]: didn't necessarily have to be repaired, I think just "block not found" works

File: wen-restart/src/last_voted_fork_slots_aggregate.rs [hunk 3]
```diff
@@ -131,15 +137,25 @@ impl LastVotedForkSlotsAggregate {
...
-        let active_percent = total_active_stake as f64 / total_stake as f64 * 100.0;
+        })
+    }
+
+    pub(crate) fn get_aggregate_result(&self) -> LastVotedForkSlotsAggregateResult {
         LastVotedForkSlotsAggregateResult {
             slots_to_repair: self.slots_to_repair.iter().cloned().collect(),
-            active_percent,
+            active_percent: self.total_active_stake() as f64
+                / self.epoch_stakes.total_stake() as f64
```
Comment (carllin) [inline wen-restart/src/last_voted_fork_slots_aggregate.rs]: This epoch stakes is dependent on the root bank that was passed in on startup right? Won't this vary if some people have roots before/after an epoch boundary?

Might be good to also log at some point which root bank/epoch we are starting at, I have a feeling it might be useful for debugging

Let's make a note to add a test case for running wen restart across epoch boundaries when everything is all checked in

File: wen-restart/src/last_voted_fork_slots_aggregate.rs [hunk 4]
```diff
@@ -131,15 +137,25 @@ impl LastVotedForkSlotsAggregate {
...
+                * 100.0,
+        }
+    }
+
+    pub(crate) fn get_final_result(&self) -> LastVotedForkSlotsFinalResult {
+        LastVotedForkSlotsFinalResult {
+            slots_stake_map: self.slots_stake_map.clone(),
+            total_active_stake: self.total_active_stake(),
```
Comment (carllin) [inline wen-restart/src/last_voted_fork_slots_aggregate.rs]: don't think this needs to be a clone? The `LastVotedForkSlotsAggregate` is never used after this so its internal `slots_stake_map` can just be moved if we just pass `self` instead of `&self`

File: wen-restart/src/last_voted_fork_slots_aggregate.rs [hunk 5]
```diff
@@ -131,15 +137,25 @@ impl LastVotedForkSlotsAggregate {
...
+        self.active_peers.iter().fold(0, |sum: u64, pubkey| {
             sum.saturating_add(Self::validator_stake(&self.epoch_stakes, pubkey))
-        });
-        let active_percent = total_active_stake as f64 / total_stake as f64 * 100.0;
+        })
+    }
+
+    pub(crate) fn get_aggregate_result(&self) -> LastVotedForkSlotsAggregateResult {
         LastVotedForkSlotsAggregateResult {
             slots_to_repair: self.slots_to_repair.iter().cloned().collect(),
```
Comment (carllin) [inline wen-restart/src/last_voted_fork_slots_aggregate.rs]: I don't think we need to clone here every time or even have a separate `LastVotedForkSlotsAggregateResult` struct. I think we can just expose:

1.  an iterator through `fn slots_to_repair(&self)` for the `last_voted_fork_slots_aggregate.slots_to_repair` 
2. an `active_percent()` function that we call


Also, let's move all of this refactoring into a separate PR

File: wen-restart/src/wen_restart.rs [hunk 6]
```diff
@@ -196,7 +232,100 @@ pub(crate) fn aggregate_restart_last_voted_fork_slots(
...
+
+pub(crate) fn find_heaviest_fork(
+    aggregate_final_result: LastVotedForkSlotsFinalResult,
+    bank_forks: Arc<RwLock<BankForks>>,
+    exit: Arc<AtomicBool>,
+) -> Result<(Slot, Hash)> {
+    // Under normal cases without duplicate blocks, we should have a single chain of
+    // blocks from the root to the newest block with > (67% - 5% -  (100% - X%)) of stake.
+    // 1. If X% of the validators joined the restart (X > 80%), then there can't be two
+    //    kids B and C of the same parent A with more than (67% - 5% -  (100% - X%)) of
```
Comment (carllin) [inline wen-restart/src/wen_restart.rs]: kids -> children

File: wen-restart/src/wen_restart.rs [hunk 7]
```diff
@@ -196,7 +232,100 @@ pub(crate) fn aggregate_restart_last_voted_fork_slots(
...
+    // In both cases, we should stop and output warning on the smallest abnormal block we find
+    // in the chain, since that is likely the root of the problem.
+    // Therefore, we will filter out all slots greater than local root with more than
+    // (active_stake_percnet - 38%) of the stake and sort them. Then for each slot, we check
+    // that the block we have:
+    // 1. Chains to the root/last block we examined
+    // 2. Has less stake than its parent
+    // Abort if any block fails above checks and return error on the block.
+
+    // Because everything else is stopped, it's okay to grab a big lock on bank_forks.
```
Comment (carllin) [inline wen-restart/src/wen_restart.rs]: I think this comment doesn't serve a good purpose.

It's too much detail for a function comment, and not enough detail to actually be a proof or justify any of the math. For instance, why `(67% - 5% - (100% - X%))`? 

We should also move this to be a function comment. We can point to the SIMD for the full proof, and just say something along the lines of

"finds the unique heaviest common ancestor by aggregating the stake weights of each slot in `aggregate_final_result`, using `bank_forks` as the structure for the tree"

File: wen-restart/src/wen_restart.rs [hunk 8]
```diff
@@ -196,7 +232,100 @@ pub(crate) fn aggregate_restart_last_voted_fork_slots(
...
+    let my_bank_forks = bank_forks.read().unwrap();
+    let root_bank = my_bank_forks.root_bank().clone();
+    let root_slot = root_bank.slot();
+    // TODO: Should use better epoch_stakes later.
```
Comment (carllin) [inline wen-restart/src/wen_restart.rs:254]: This seems true in multiple places, should create an issue for this so we don't forget

File: wen-restart/src/wen_restart.rs [hunk 9]
```diff
@@ -196,7 +232,100 @@ pub(crate) fn aggregate_restart_last_voted_fork_slots(
...
+    aggregate_final_result: LastVotedForkSlotsFinalResult,
+    bank_forks: Arc<RwLock<BankForks>>,
+    exit: Arc<AtomicBool>,
+) -> Result<(Slot, Hash)> {
+    // Under normal cases without duplicate blocks, we should have a single chain of
+    // blocks from the root to the newest block with > (67% - 5% -  (100% - X%)) of stake.
+    // 1. If X% of the validators joined the restart (X > 80%), then there can't be two
+    //    kids B and C of the same parent A with more than (67% - 5% -  (100% - X%)) of
+    //    the stake each. Assume B and C exists, their combined stake is more than
+    //    2 * (67% - 5% - (100% - X%)) = 134% - 10% - 2 * (100% - X%) = 2*X% - 76% > X%.
```
Comment (carllin) [inline wen-restart/src/wen_restart.rs]: Well somebody could double vote maliciously, which means the combined stake would double count them, which would break this proof

I think the proof needs the additional constraint that the beginning split for a fork has to be at least 47-33, and cannot be something like 40-40

File: wen-restart/src/wen_restart.rs [hunk 10]
```diff
@@ -196,7 +232,100 @@ pub(crate) fn aggregate_restart_last_voted_fork_slots(
...
+    //    Since we can only have X% of the stake, this is a contradiction.
+    // 2. If there are no duplicate blocks and a child got Y% of the stake, then its
+    //    parent should at least have Y% of the stake.
+    // So, we can start from the latest block with > (67% - 5% -  (100% - X%)) = X% - 38%
```
Comment (carllin) [inline wen-restart/src/wen_restart.rs]: Hmm, I don't think such a child necessarily exists?

Say a validator V contributed to voting on the last optimistically confirmed slot X.  They could have then voted on some other  last slot that Y>X such that Y does not have 42% of the stake

File: wen-restart/src/wen_restart.rs (hunk 1 above)
Comment (wen-coding) [reply wen-restart/src/wen_restart.rs]: Not necessarily, we may want to adjust switching proof left and right for other reasons, but as long as 67% for OC and 5% for malicious doesn't change, we should have (67% - 5% - (100% - active_stake)) = active_stake - 38% here.
I'll update the comments.

File: wen-restart/src/wen_restart.rs (hunk 2 above)
Comment (wen-coding) [reply wen-restart/src/wen_restart.rs]: Done.

File: wen-restart/src/last_voted_fork_slots_aggregate.rs (hunk 3 above)
Comment (wen-coding) [reply wen-restart/src/last_voted_fork_slots_aggregate.rs]: The whole stake thing is in a bad shape now, I plan to put the use correct epoch stake into another PR. Added a todo here.

File: wen-restart/src/last_voted_fork_slots_aggregate.rs (hunk 4 above)
Comment (wen-coding) [reply wen-restart/src/last_voted_fork_slots_aggregate.rs]: Done.

File: wen-restart/src/wen_restart.rs (hunk 8 above)
Comment (wen-coding) [reply wen-restart/src/wen_restart.rs:254]: Sure, https://github.com/anza-xyz/agave/issues/229 filed.

File: wen-restart/src/wen_restart.rs (hunk 7 above)
Comment (wen-coding) [reply wen-restart/src/wen_restart.rs]: Changed and moved the whole discussion to SIMD.

File: wen-restart/src/wen_restart.rs (hunk 6 above)
Comment (wen-coding) [reply wen-restart/src/wen_restart.rs]: Moved the whole discussion to SIMD.

File: wen-restart/src/wen_restart.rs (hunk 9 above)
Comment (wen-coding) [reply wen-restart/src/wen_restart.rs]: Moved the whole discussion to SIMD.

File: wen-restart/src/wen_restart.rs (hunk 10 above)
Comment (wen-coding) [reply wen-restart/src/wen_restart.rs]: Moved the whole discussion to SIMD.

File: wen-restart/src/last_voted_fork_slots_aggregate.rs (hunk 5 above)
Comment (wen-coding) [reply wen-restart/src/last_voted_fork_slots_aggregate.rs]: Actually it's not a big change, so I've refactored it.

Comment (carllin) [reply wen-restart/src/last_voted_fork_slots_aggregate.rs]: Functionally these changes along with the blockstore plumbing just don't belong in the this PR. It makes it hard to revert/organize in the future. I would extract them.

Comment (wen-coding) [reply wen-restart/src/last_voted_fork_slots_aggregate.rs]: Sure, I rolled back all these changes.

Comment (carllin) [reply wen-restart/src/last_voted_fork_slots_aggregate.rs]: we should land those refactoring and plumbing changes before this one

File: wen-restart/src/wen_restart.rs [hunk 11]
```diff
@@ -263,46 +264,31 @@ pub(crate) fn find_heaviest_fork(
...
         if exit.load(Ordering::Relaxed) {
             return Err(WenRestartError::Exiting.into());
         }
-        if let Some(new_bank) = my_bank_forks.get(slot) {
-            if new_bank.parent_slot() != current_bank.slot() {
+        if let Ok(Some(block_meta)) = blockstore.meta(slot) {
+            info!("block meta: {:?}", block_meta);
+            if block_meta.parent_slot != Some(expected_parent) {
+                info!(
+                    "Block {} is not linked to expected parent {} but to {:?}",
```
Comment (carllin) [inline wen-restart/src/wen_restart.rs]: nit: `Block in blockstore is not linked to expected parent from Wen Restart {} but to `

File: wen-restart/src/wen_restart.rs [hunk 12]
```diff
@@ -198,7 +235,60 @@ pub(crate) fn aggregate_restart_last_voted_fork_slots(
...
+        .collect::<Vec<Slot>>();
+    slots.sort();
+    let mut expected_parent = root_slot;
+    for slot in slots {
+        if exit.load(Ordering::Relaxed) {
+            return Err(WenRestartError::Exiting.into());
+        }
+        if let Ok(Some(block_meta)) = blockstore.meta(slot) {
+            info!("block meta: {:?}", block_meta);
+            if block_meta.parent_slot != Some(expected_parent) {
```
Comment (carllin) [inline wen-restart/src/wen_restart.rs:273]: Hmmm if you start with a really old root `N` and everyone else's root is at `N + 1000`, are you guaranteed to have gotten all the slots between your root and everyone else's root through the aggregation phase? 

Seems like it's not guaranteed if people didn't gossip the older slots in their last voted forks

File: wen-restart/src/wen_restart.rs (hunk 11 above)
Comment (wen-coding) [reply wen-restart/src/wen_restart.rs]: Done.

File: wen-restart/src/wen_restart.rs (hunk 12 above)
Comment (wen-coding) [reply wen-restart/src/wen_restart.rs:273]: We always send back (last_vote - 64k, last_vote), it may include the local root and slots older than local root. So yes you should see N to N+1000 in the RestartLastVotedForkSlots and you will aggregate those.

The number of slots may be less than 64k, depending on how much we can fit in one UDP packet, but it's definitely more than 1000.

Comment (carllin) [reply wen-restart/src/wen_restart.rs:273]: Mayeb a custom error message would be good here if the only mismatch is on the root? Say that this means potentially we don't have enough ancestors or your root is too old?

Comment (wen-coding) [reply wen-restart/src/wen_restart.rs:273]: Sure, I think an error log on first mismatch being on root is useful. But not sure we need the error log when the root mismatch is the only error, because even if the list is screwed up we might want to print root mismatch, changed.

**Review by carllin — APPROVED** [review]
